### DISCUSSION TITLE
Regarding GraalVM CE and Java performance comparison

### DISCUSSION THREAD

**Discussion by Roopsai507** [oracle/graal#5680] [Q&A] [answered]
Hi, I have made my application run with graalvm java. I benchmarked the performance and saw that openJDK-java is faster than the graalvm java. And also memory footprint of my application when I run with java is around 60 MB whereas in graalVM CE it was about 130MB. ( My application consists of running a jetty server and when a request is received by the server, it will dynamically load a class and send the response.)
These are the benchmark results for java and graal
```
Running 10m test @ https://11.255.15.161:8443
  8 threads and 12 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    12.57ms   48.56ms   1.89s    97.49%
    Req/Sec   106.65     71.12   333.00     60.47%
  63609 requests in 10.00m, 17.47MB read
  Socket errors: connect 0, read 0, write 0, timeout 269
Requests/sec:    106.00
Transfer/sec:     29.81KB

JAVA
Running 10m test @ https://11.255.15.161:8443
  8 threads and 12 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    31.64ms  108.84ms   1.98s    93.84%
    Req/Sec    93.23    157.74     0.96k    86.62%
  92170 requests in 10.00m, 25.31MB read
  Socket errors: connect 0, read 0, write 0, timeout 104
Requests/sec:    153.59
Transfer/sec:     43.19KB

NATIVE_IMAGE
Running 10m test @ https://11.255.15.161:8443
  8 threads and 16 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    64.86ms  167.74ms   1.99s    93.02%
    Req/Sec    58.92     48.52   360.00     63.29%
  88837 requests in 10.00m, 24.39MB read
  Socket errors: connect 0, read 0, write 0, timeout 556
Requests/sec:    148.03
Transfer/sec:     41.63KB
```

I have also tested it by converting the application to native-image (memory footprint of binary about 50MB). As per the above stats, java is faster than graal and native-image. Is it the expected performance/behavior with GraalVM Community Edition? Will GraalVM Enterprise Edition gives faster performance ? If so, what is the use of the GraalVm community edition's java and native-image ??

**Comment by alina-yur** [comment, marked as answer]
Hi @Roopsai507, thank you for your question. 

To comment on specific performance characteristics, we would need more details about the benchmark setup. This includes the benchmark available in source code form, a description of the target machine, a script for producing the load, and a description of the tools and measurement methodologies applied.

For best throughput peak performance, we do recommend the GraalVM Enterprise Edition. You can use it for free for development and evaluation purposes; it's also free on Oracle Cloud. You can download it [here](https://www.oracle.com/downloads/graalvm-downloads.html). For a detailed evaluation on different workloads of GraalVM CE, EE, and OpenJDK, see the Renaissance benchmark suite at https://renaissance.dev.

Native Image is typically recommended when startup and memory footprint matters. It can also achieve high throughput, but primarily when used in connection with [profile-guided optimizations](https://www.graalvm.org/22.3/reference-manual/native-image/guides/optimize-native-executable-with-pgo/) available in the Enterprise Edition.
